### ISSUE TITLE
Problems loading Xgboost models

### ISSUE THREAD

**Issue by selphie1410** [dmlc/xgboost#3696]
I have trained an Xgboost regression model in R and I want to use it other data, Sadly I cannot convert it to PMML as it is a regression model.

I have made a test set which is a csv file that I want to use my model on.

I read in the test set and run  my model.
.
I then want to save my model load it and run it on the same data. When I do this I get totally different answers on the same data

Original
 1.1608219 1.1608219 0.7352632 0.7352632 0.8696429 1.0045016 0.8416470 1.0045016
after reloading
3.487421 3.487421 3.487421 3.487421 3.487421 3.487421 2.712625 3.487421

I have used xgb.save and xgb.load

I have also used the .bin hack in another issue here. Didn't work but another different set of answers.

Obviously something is wrong but I can't work out what.

Any other ideas how to deploy an xgboost regression model on data without retraining it each time?

**Comment by hcho3** [comment]
Can you post your R script so that I can try running it?

**Comment by selphie1410** [comment]
testsheetcsv <- read.csv("~/2018 racing models/testsheetcsv.csv")
 View(testsheetcsv)
data=testsheetcsv
AW12.XG <- read.csv("~/2018 racing models/AW12 XG.csv")
trainIndex   <- createDataPartition(AW12.XG$Dist12, p=0.75, list=FALSE, times=1)
All_rmse<- c()
 Param_group<-c()
 params <- list(booster = "gbtree", objective = "reg:linear", eta=0.3, gamma=0, max_depth=6, min_child_weight=1, subsample=1, colsample_bytree=1)
 xgbcv <- xgb.cv( params = params, data = dtrain, nrounds = 100, nfold = 5, showsd = T, stratified = T, print.every.n = 10, early.stop.round = 20, maximize = F)
#just get a quick best model
xgb1 <- xgb.train (params = params, data = dtrain, nrounds = 8, watchlist = list(val=dtest,train=dtrain), print.every.n = 10, early.stop.round = 10, maximize = F , eval_metric = "error")
predict(xgb1,newdata=as.matrix(data[, -12]))
[1] 1.1608219 1.1608219 0.7352632 0.7352632 0.8696429 1.0045016 0.8416470 1.0045016
xgb.save(xgb1, "Awmodel")
[1] TRUE
awmod1 <- xgb.load("Awmodel")
 predict(awmod1,newdata=as.matrix(data[, -12]))
[1] 2.359888 2.650479 1.321146 0.838798 3.043278 2.277152 1.482323 2.767058
#not the same?

**Comment by selphie1410** [comment]
No idea why its scored out lines???

**Comment by hcho3** [comment]
Is the dataset you're using public?

**Comment by selphie1410** [comment]
no. I was hoping it wouldn’t be a problem,
It consists of 8 0/1 variables and three real numbers.

Using R studio to import them from CSV files.

**Comment by hcho3** [comment]
@selphie1410 

> I was hoping it wouldn’t be a problem

Without the data, diagnosis of the bug is difficult. Can you generate a random dataset that looks like your dataset?

**Comment by selphie1410** [comment]
I can let you have the originals. I’ll put them up shortly.

**Comment by selphie1410** [comment]
[GIT Stuff.xlsx](https://github.com/dmlc/xgboost/files/2378046/GIT.Stuff.xlsx)
Here you are

**Comment by hcho3** [comment]
Thanks, I tried running your script and get the error 
```
Error in terms.formula(object, data = data) : 
  'data' argument is of the wrong type
Calls: sparse.model.matrix -> terms -> terms.formula
Execution halted
```

**Comment by selphie1410** [comment]
I don’t get that! 

I can build the model quite happily. 

All the spare.model.matrix does is convert the CSV file into something XGb.Dmatrix can handle.

What do you need me to do?

**Comment by hcho3** [comment]
Can you post the CSV files you converted from XLS?

**Comment by selphie1410** [comment]
[testcsv.zip](https://github.com/dmlc/xgboost/files/2380283/testcsv.zip)

This is the CSV file zipped up because GIThub doesn't let me upload csv

**Comment by hcho3** [comment]
@selphie1410 testcsv.zip doesn't contain Dist12. Did you upload the right file?

**Comment by hcho3** [comment]
@selphie1410 There are several issues with the code you posted:
1. The variables `train` and `test` were not defined. I had to add
```R
train <- AW12.XG[trainIndex,]
test <- AW12.XG[-trainIndex,]
```
2. The variables `train.label` and `test.label` were not defined either. Which column is supposed to be the target? Is `Dist12` column the label?

**Comment by hcho3** [comment]
@selphie1410 I found the problem: since you are using early stopping, you have to specify `ntreelimit` when predicting:
```R
predict(xgb1,newdata=as.matrix(data[, -12]))
xgb.save(xgb1, "Awmodel")
awmod1 <- xgb.load("Awmodel")
predict(awmod1,newdata=as.matrix(data[, -12]), ntreelimit=1)   # 1 was the best iteration for xgb1
```
When you save a model and load it back, the best iteration isn't saved. To preserve early stopping information, you should save the R object using `saveRDS()`.
